Tests/NumLib/TestExtrapolation.cpp

```diff
@@ -0,0 +1,359 @@
+        GlobalSetup::transform(
+                local_asm_builder,
+                mesh.getElements(),
+                _local_assemblers,
+                _integration_order);
+    }
+
+    static unsigned const _integration_order = 2;
+
+    MeshLib::MeshSubset _mesh_subset_all_nodes;
+    std::unique_ptr<AssemblerLib::LocalToGlobalIndexMap> _dof_table;
+
+    std::unique_ptr<GlobalAssembler> _global_assembler;
+    std::vector<std::unique_ptr<LocalAssembler>> _local_assemblers;
+
+    std::unique_ptr<ExtrapolatorInterface> _extrapolator;
+};
+
+template<typename GlobalSetup>
+unsigned const TestProcess<GlobalSetup>::_integration_order;
+
+
+template<typename GlobalSetup>
+void extrapolate(TestProcess<GlobalSetup> const& pcs,
+                 IntegrationPointValue property,
+                 typename GlobalSetup::VectorType const&
+                 expected_extrapolated_global_nodal_values,
+                 std::size_t const nnodes, std::size_t const nelements)
+{
+    namespace BLAS = MathLib::BLAS;
+    using GlobalVector = typename GlobalSetup::VectorType;
+
+    auto const tolerance_res = std::numeric_limits<double>::epsilon();
+    auto const tolerance_dx  = 20.0 * std::numeric_limits<double>::epsilon();
+
+    auto const  result   = pcs.extrapolate(property);
+    auto const& x_extra  = *result.first;
+    auto const& residual = *result.second;
+
+    ASSERT_EQ(nnodes,    x_extra.size());
+    ASSERT_EQ(nelements, residual.size());
+
+    auto const res_norm = BLAS::normMax(residual);
+    DBUG("maximum norm of residual: %g", res_norm);
+    EXPECT_GT(tolerance_res, res_norm);
+
+    auto delta_x = MathLib::MatrixVectorTraits<GlobalVector>::newInstance(
+                expected_extrapolated_global_nodal_values);
+    BLAS::axpy(*delta_x, -1.0, x_extra); // delta_x = x_expected - x_extra
+
+    auto const dx_norm = BLAS::normMax(*delta_x);
+    DBUG("maximum norm of delta x:  %g", dx_norm);
+    EXPECT_GT(tolerance_dx, dx_norm);
+}
+
+#ifndef USE_PETSC
+TEST(NumLib, Extrapolation)
+#else
+TEST(NumLib, DISABLED_Extrapolation)
+#endif
+{
+    /* In this test a random vector x of nodal values is created.
+     * This vector is interpolated to the integration points using each
+     * element's the shape functions.
+     * Afterwards the integration point values y are extrapolated to the mesh
+     * nodes again.
+     * Since y have been obtained from x via interpolation, it is expected, that
+     * the interpolation result nearly exactly matches the original nodal values
+     * x.
+     */
+
+    namespace BLAS = MathLib::BLAS;
+    using GlobalSetup = GlobalSetupType;
+    using GlobalVector = GlobalSetup::VectorType;
+
+    const double mesh_length = 1.0;
+    const double mesh_elements_in_each_direction = 5.0;
+
+    // generate mesh
+    std::unique_ptr<MeshLib::Mesh> mesh(
+                MeshLib::MeshGenerator::generateRegularHexMesh(
+                    mesh_length, mesh_elements_in_each_direction));
```

Comment:
TODO program a general mesh generator for all mesh element types and shape fcts.
In this test: + multiple integration orders.